var/spack/repos/builtin/packages/python/package.py

```diff
@@ -945,10 +945,27 @@ def setup_dependent_build_environment(self, env, dependent_spec):
         if not is_system_path(path):
             env.prepend_path('PATH', path)
 
-        for d in dependent_spec.traverse(deptype=('build', 'run', 'test'), root=True):
-            if d.package.extends(self.spec):
-                env.prepend_path('PYTHONPATH', join_path(
-                    d.prefix, self.site_packages_dir))
+        # Add installation prefix to PYTHONPATH, needed to run import tests
+        if dependent_spec.package.extends(self.spec):
+            env.prepend_path(
+                'PYTHONPATH', join_path(dependent_spec.prefix, self.site_packages_dir)
+            )
+
+        for direct_dep in dependent_spec.dependencies(deptype=('build', 'run', 'test')):
+            # Add direct build/run/test dependencies to PYTHONPATH, needed to build
+            # the package and to run import tests
+            if direct_dep.package.extends(self.spec):
+                env.prepend_path(
+                    'PYTHONPATH', join_path(direct_dep.prefix, self.site_packages_dir)
+                )
+                # Add recursive run dependencies of all direct dependencies,
+                # needed by direct dependencies at run-time
+                for indirect_dep in direct_dep.traverse(deptype='run'):
+                    if indirect_dep.package.extends(self.spec):
+                        env.prepend_path(
+                            'PYTHONPATH',
+                            join_path(indirect_dep.prefix, self.site_packages_dir),
+                        )
```

Comment:
@tgamblin @alalazo @becker33 @scheibelp can you look over this real quick and make sure this looks correct? Previously, we were adding build deps of build deps to the `PYTHONPATH`, so packages that were missing deps were still building correctly. I _think_ the correct behavior should be to add:

1. The package itself
2. Any direct build/run/test deps
3. Any indirect run deps

Does this logic seem correct? Does this code seem correct for implementing this logic?

Reply:
Also, this likely adds a lot of duplicate entries to `PYTHONPATH`. I assume `prepend_path` doesn't have any special handling for this. I'll convert this to a set later.

Reply:
The package should only set variables for the package itself, it should never recurse.

Reply:
As discussed below, this currently doesn't work.